### PR TITLE
Enable gradient scaling in autograd

### PR DESCRIPTION
**Pull request by mcarilli**
## Context

The dream is for Pytorch to support automatic mixed precision when run in eager mode, jit mode, or a combination.  I have three stages planned:
1. Gradient scaling
2. a. (for eagerly-executed functions) Patching of automatic whitelist-blacklist-based casts into the autogenerated `CUDAType.cpp` functions (or wherever you all think is appropriate). (@zdevito @mruberry i totally agree that putting optional casts in the VariableType* code path, where they would affect everything, was a bad idea, but I think restricting them to the code path for Cuda tensors alone seems reasonable, since that's where they are necessary.  We can talk more later...)
b.  (for jitted functions) Creation of a jit pass to handle whitelist-blacklist casting for fusion groups (@jjsjann123 is point person here).
3. Figuring out the user-facing API.

I think each of these can be treated as modular, independent PRs.  Also, the current PR is **not intended to create anything user-facing**, it's intended to provide the tools we will need to create reasonable implementations for 2 and 3.  I don't believe the exposure of gradient scaling needs to be complex to support a wide range of possible implementations for 2 and 3.  I also believe the simple gradient scaling API provided by the current PR will be sufficient for 2 and 3, certainly for the plans I have in mind.  Again, for the moment nothing is intended to be user-facing, so I'm totally OK with the current PR being a "try it on for size" situation which may change in the future.  I'm also OK with this PR serving as the tracker for a monolithic, as opposed to modular, integration of 1, 2, and 3 all together.

## This PR

The current PR is my first attempt at enabling gradient scaling in a way that
- is compatible with both eager mode and jitted functions
- is functionally correct for all typical single and double backward use cases
- works for backward passes invoked from either Python or C++ user scripts
- exposes the necessary+sufficient levers to work with a future full automatic mixed precision integration.

As far as I can tell, gradients only enter the autograd engine through the argument list of [engine.execute](https://github.com/pytorch/pytorch/compare/master...mcarilli:loss_scaling_rebase?expand=1#diff-01f226b610655c49b1e52eddf9ea4126R694), and are only returned to the user in one of two ways:
1.  If an ordinary `loss.backward` or `torch.autograd.backward` is underway, they are consumed by leaf `AccumulateGrad` functions to populate `.grad` attributes.  
2. If a `torch.autograd.grad` call is underway to produce gradients out-of-place (e.g. to set up a gradient penalty term) gradients are intercepted before they reach `AccumulateGrad` functions and returned as `captured_vars_`.

The core of my strategy is three modified lines of `Engine::execute` in engine.cpp that
[optionally scale incoming (root) gradients](https://github.com/pytorch/pytorch/compare/master...mcarilli:loss_scaling_rebase?expand=1#diff-01f226b610655c49b1e52eddf9ea4126R706),
[optionally unscale gradients about to be consumed by a leaf's AccumulateGrad](https://github.com/pytorch/pytorch/compare/master...mcarilli:loss_scaling_rebase?expand=1#diff-01f226b610655c49b1e52eddf9ea4126R518-R519), and
[optionally unscale gradients about to be returned out-of-place as captured_vars_](https://github.com/pytorch/pytorch/compare/master...mcarilli:loss_scaling_rebase?expand=1#diff-01f226b610655c49b1e52eddf9ea4126R775).
If a `torch.autograd.grad(..., only_inputs=False)` is underway, some gradients may be destined for `AccumulateGrad` functions and others may be returned as `captured_vars_`, but that is also handled by my approach, and covered by my tests.

Aside from those three lines, the rest is basically boilerplate Python bindings and a helper kernel to do unscaling and inf/nan checking in one pass.

I've included thorough tests for all the reasonable single and double backward use cases I could think of, and some unreasonable ones.

## Things I already know I'd like to discuss

- [ ] **Jit compatibility** After discussion with @jjsjann123 my understanding is that fusion groups are executed as functions by the autograd engine.  If so, this PR's scaling and unscaling logic will wrap around them, and everything will "just work," _unless_ a fusion group can include/subsume a leaf's AccumulateGrad function.  I don't expect this is the case (because AccumulateGrad functions don't have a direct analogue in the forward pass, and also wouldn't be written explicitly as ops in user-defined TorchScript) but I definitely want to know.
- [ ] **General organization** As I said, most of this PR is essentially boilerplate.  The organization could certainly be tightened to fewer files, but I wanted to PR it in a somewhat spread out configuration (based on `grad_mode`) and then discuss how it could be pruned.  In principle I don't even need to expose Python bindings through torch._C, I could route everything from at::native, but to my eyes `grad_mode` could have done so as well, and I assume there's a reason it didn't.  
- [ ] **Passing vectors of tensors around in engine.cpp** There are several places in engine.cpp where I wasn't sure if the best approach was to pass a vector of tensors by value, reference, or rvalue reference.  I've commented them.  Any second-guessing is appreciated.
- [ ] **C++ API tests** I tried to write thorough Python test coverage, and I also ran some C++-based tests locally, but I've never added pure C++ tests to Pytorch before.  I can keep digging on my own, but any pointers in the right direction would be appreciated.

### CODE REVIEW

**Comment by mcarilli** [comment]
Closing because a Python-side approach to gradient scaling may be more flexible and transparent.
